pkg/template/controller/templateinstance_finalizer.go

```diff
@@ -0,0 +1,261 @@
+// linked Secret object.  The TemplateInstanceFinalizerController instantiates objects
+// using its own service account, first verifying that the requester also has
+// permissions to instantiate.
+type TemplateInstanceFinalizerController struct {
+	dynamicRestMapper meta.RESTMapper
+	client            dynamic.DynamicInterface
+	jsonConfig        *rest.Config
+	templateClient    templateclient.Interface
+
+	lister         templatelister.TemplateInstanceLister
+	informerSynced func() bool
+
+	queue workqueue.RateLimitingInterface
+
+	readinessLimiter workqueue.RateLimiter
+
+	clock clock.Clock
+}
+
+// NewTemplateInstanceFinalizerController returns a new TemplateInstanceFinalizerController.
+func NewTemplateInstanceFinalizerController(dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper, config *rest.Config, templateClient templateclient.Interface, informer internalversion.TemplateInstanceInformer) *TemplateInstanceFinalizerController {
+	c := &TemplateInstanceFinalizerController{
+		dynamicRestMapper: dynamicRestMapper,
+		templateClient:    templateClient,
+		lister:            informer.Lister(),
+		informerSynced:    informer.Informer().HasSynced,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
+		readinessLimiter:  workqueue.NewItemFastSlowRateLimiter(5*time.Second, 20*time.Second, 200),
+		clock:             clock.RealClock{},
+	}
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
+	c.jsonConfig.ContentConfig = dynamic.ContentConfig()
+	var err error
+	c.client, err = dynamic.NewForConfig(c.jsonConfig)
+	if err != nil {
+		glog.Errorf("failure creating dynamic client: %v", err)
+		return nil
+	}
+
+	return c
+}
+
+// getTemplateInstance returns the TemplateInstance from the shared informer,
+// given its key (dequeued from c.queue).
+func (c *TemplateInstanceFinalizerController) getTemplateInstance(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.lister.TemplateInstances(namespace).Get(name)
+}
+
+// sync is the actual controller worker function.
+func (c *TemplateInstanceFinalizerController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	needsFinalizing := false
+	for _, v := range templateInstance.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer && templateInstance.DeletionTimestamp != nil {
```

Comment:
move the deletiontimestamp check out of the for loop.  Just check above and return.

Reply:
fixing